cpp/include/raft/core/bitset.cuh

```diff
@@ -208,100 +233,197 @@ struct bitset {
+        if constexpr (bitset_element_size == 64) {  // Needed because __popc doesn't support 64bit
+          if (index == n_elements_ - 1)
+            result = index_t(raft::detail::native_popc<uint64_t>(element & last_element_mask));
+          else
+            result = index_t(raft::detail::native_popc<uint64_t>(element));
+        } else {
+          if (index == n_elements_ - 1)
+            result = index_t(__popc(element & last_element_mask));
+          else
+            result = index_t(__popc(element));
+        }
+
+        return result;
+      });
+  }
+  /**
+   * @brief Returns the number of bits set to true.
+   *
+   * @return index_t Number of bits set to true
+   */
+  auto count() -> index_t
+  {
+    auto count_gpu_scalar = raft::make_device_scalar<index_t>(res_, 0.0);
+    count(count_gpu_scalar.view());
+    index_t count_cpu = 0;
+    raft::update_host(
+      &count_cpu, count_gpu_scalar.data_handle(), 1, resource::get_cuda_stream(res_));
+    resource::sync_stream(res_);
+    return count_cpu;
+  }
+  /**
+   * @brief Checks if any of the bits are set to true in the bitset.
+   */
+  bool any() { return count() > 0; }
+  /**
+   * @brief Checks if all of the bits are set to true in the bitset.
+   */
+  bool all() { return count() == bitset_len_; }
+  /**
+   * @brief Checks if none of the bits are set to true in the bitset.
+   */
+  bool none() { return count() == 0; }
+
+  bitset<bitset_t, index_t>& operator|=(const bitset<bitset_t, index_t>& other)
+  {
+    RAFT_EXPECTS(size() == other.size(), "Sizes must be equal");
+    auto this_span  = to_mdspan();
+    auto other_span = other.to_mdspan();
+    raft::linalg::map(
+      res_,
+      this_span,
+      [] __device__(bitset_t this_element, bitset_t other_element) {
+        return this_element | other_element;
+      },
+      raft::make_const_mdspan(this_span),
+      other_span);
+    return *this;
+  }
+  bitset<bitset_t, index_t>& operator&=(const bitset<bitset_t, index_t>& other)
+  {
+    RAFT_EXPECTS(size() == other.size(), "Sizes must be equal");
+    auto this_span  = to_mdspan();
+    auto other_span = other.to_mdspan();
+    raft::linalg::map(
+      res_,
+      this_span,
+      [] __device__(bitset_t this_element, bitset_t other_element) {
+        return this_element & other_element;
+      },
+      raft::make_const_mdspan(this_span),
+      other_span);
+    return *this;
+  }
+  bitset<bitset_t, index_t>& operator^=(const bitset<bitset_t, index_t>& other)
+  {
+    RAFT_EXPECTS(size() == other.size(), "Sizes must be equal");
+    auto this_span  = to_mdspan();
+    auto other_span = other.to_mdspan();
+    raft::linalg::map(
+      res_,
+      this_span,
+      [] __device__(bitset_t this_element, bitset_t other_element) {
+        return this_element ^ other_element;
+      },
+      raft::make_const_mdspan(this_span),
+      other_span);
+    return *this;
   }
 
  private:
   raft::device_uvector<bitset_t> bitset_;
```

Comment:
This is buried in the container policy code for RAFT and I'd prefer using RAFT types, even if they are just direct wrappers around thrust or rmm types (it  provides us a safe facade to maintain api compatibility even if the underlying impls should somehow change or need to be modified).

Reply:
raft::device_vector is the way to go here!

Reply:
Not really, I chose `raft::device_uvector` because it is resizable. `raft::device_vector`  isn't, and this resizable feature is very helpful for incremental addition to IVF indexes (and soon CAGRA)

Reply:
For my own education, won't `raft::device_vector` initialize the underlying memory? Would it be worth exposing `raft::device_uvector` more publicly specifically for cases like this?

Reply:
`resize()` was indeed a design decision we originally took because of `stream` parameter support for device mdarrays vs host mdarrays not needing one. However, we accept `raft::resources` in the mdarray constructor, I wonder if it would be okay to `resize()` with that stream @cjnolet ?

@wphicks AFAIK mdarray does not initialize the memory and I couldn't locate any code for the same. Could you point to where that is happening?

Reply:
@divyegala You're quite right! Had `rmm::device_vector` in my mind instead of `raft::device_vector`. `raft::device_vector` uses `rmm::device_uvector` for its container policy, so no initialization.

Reply:
Adding my two cents on the resize question, I'd be quite squeamish about using the resources from the constructor. Not only do we not know if that resources object has gone out of scope, but it also makes it much harder to track and ensure stream safety if we are not explicitly passing the stream we want to use when a resize might occur. For example:
```
auto res1 = device_resources{};  // Some underlying stream
auto arr = device_mdarray{res1, ...};
res1.sync_stream();
auto res2 = device_resources();
// modify underlying data of arr using the stream from res2
res2.sync_stream();
// Is it now safe to copy back data to the host on res2? Maybe. If a resize was triggered in between, we'd need another res1.sync_stream() first
```
Alternatively, if we sync the stream from res1 in the resize method, we might still have an issue if the data are actively being modified on another stream during the resize.

Reply:
@wphicks thanks for typing that out, you are right and that just clarifies me for again why we didn't end up supporting this in the first place anyway.

`raft::device_uvector` is supposed to be an internal/detail type and I do not belive it should be used directly. If `resize()` is needed then we should switch the type to `rmm::device_uvector`. Thoughts on this @cjnolet @wphicks ?

Reply:
Actually, the original suggestion to use a `raft` type instead of directly using the rmm type in public API code was from me. To make the point more clear, this is about consistency and safeguarding our implementations from changes that we cannot control. This is in a very similar vein to our diligent wrapping of the CUDA math libraries APIs so that we can centralize those calls and change any underlying details should the rug get pulled out from under us. 

That being said, I would actually suggest that we not implement resize at all. The reason being that when we grow one of the existing `raft_device_matrix` buffers in IVF-Flat and IVF-pq, we're currently reallocating anyways, and this is also what the `device_uvector` is doing under the hood. 

As far as reusing resources- I don't think it's bad to assume that a resource going out of scope/becoming invalid means that all objects which use those resources become invalid anyways. It's the user's responsiblity to make sure that any and all resources created and propagated through the APIs with a `raft::device_resources` instance should be properly maintained throughout the lifecycle of the calls they make into the RAFT APIs. These resources are meant to be held on the stack, and as such, these calls are meant to be stateless, which means they should complete before the resources are invalidated.

Reply:
I removed raft::resources instance from the bitset class.